jedi/parser/tree.py

```diff
@@ -52,6 +53,19 @@ def is_node(node, *symbol_names):
         return type in symbol_names
 
 
+def _fix_forward_reference(annotation):
+    if isinstance(annotation, String):
```

Comment:
Well, the PEP is vague on whether they should be allowed. Right under your quoted line, it says:

> Moreover, the expression should be parseable as a valid type hint, i.e., it is constrained by the rules from the section Acceptable type hints above.

The Acceptable type hints section says:

> Annotations should be kept simple or static analysis tools may not be able to interpret the values. For example, dynamically computed types are unlikely to be understood. (This is an intentionally somewhat vague requirement, specific inclusions and exclusions may be added to future versions of this PEP as warranted by the discussion.)

Having said that, if this is supported (more or less out of the box) by jedi, no reason not to do it (and test it). See new commit. However I don't think jedi is smart enough to realise that your `x` is out of scope (also see new commit), unless there's something I'm not getting.